### PR TITLE
[logging] E2E tests for logging playbook

### PR DESCRIPTION
**Pull request by suhanime**
This commit add the tests to ensure Ruby and Fluentd are installed correctly
The tests are meant to run on the Windows instance after the logging
playbook has been run successfully against it

### CODE REVIEW

File: tools/ansible/tasks/logging/test/logging_test.go [hunk 1]
```diff
@@ -1,18 +1,26 @@
...
 	"strings"
 	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
+// Pining down version of Ruby Devkit to 2.5, 64 bit version
+// Ruby installation path kept default C:\
+var rubyExecutablePath = "C:\\Ruby25-x64\\bin\\ruby.exe"
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/logging/test/logging_test.go]: Why is this needed?

File: tools/ansible/tasks/logging/test/logging_test.go [hunk 2]
```diff
@@ -37,7 +55,9 @@ func TestPluginsInstalled(t *testing.T) {
 		installedVersion, ok := installedPlugins[pluginName]
 		assert.Truef(t, ok, "Missing plugin %s", pluginName)
 		if pluginVersion != "" {
-			assert.Equalf(t, pluginVersion, installedVersion,
+			// Getting only the version in the format 1.5 instead of 1.5.6, because installation automatically
+			// fetches the latest subversion
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/logging/test/logging_test.go]: I did not understand what you mean by this? Perhaps an example? Where are you pinning the version?

File: tools/ansible/tasks/logging/test/logging_test.go [hunk 3]
```diff
@@ -25,7 +33,17 @@ var requiredPluginList = map[string]string{
...
 }
 
-// TestPluginsInstalled tests that the fluentd plugins were all installed correctly.
+// TestRubyInstalled tests if Ruby is installed correctly
+// This test should be run on a node after the logging playbook has been run against it
+// exec.Command requires fully qualified path for ruby.exe
+func TestRubyInstalled(t *testing.T) {
+	out, err := exec.Command(rubyExecutablePath, "--version").Output()
+	require.NoError(t, err)
+	actualVersion := strings.Split(string(out), " ")[1][:5]
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/logging/test/logging_test.go]: Example format again.

File: tools/ansible/tasks/logging/test/logging_test.go [hunk 4]
```diff
@@ -1,18 +1,26 @@
...
+var rubyExecutablePath = "C:\\Ruby25-x64\\bin\\ruby.exe"
+var rubyVersion = "2.5.7"
+
 // Making this a map so that if we want a specific version, the value can be the version
 var requiredPluginList = map[string]string{
-	"elasticsearch-transport":                  "",
-	"elasticsearch-api":                        "",
-	"elasticsearch":                            "",
+	// Pining down version of Fluentd to 1.5
+	"fluentd":                 "1.5",
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/logging/test/logging_test.go]: Why is fluentd plugin versioning in a different format than other plugins?

File: tools/ansible/tasks/logging/test/logging_test.go [hunk 5]
```diff
@@ -1,18 +1,26 @@
...
 	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
+// Pining down version of Ruby Devkit to 2.5, 64 bit version
+// Ruby installation path kept default C:\
+var rubyExecutablePath = "C:\\Ruby25-x64\\bin\\ruby.exe"
+var rubyVersion = "2.5.7"
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/logging/test/logging_test.go]: why is this a global var?

File: tools/ansible/tasks/logging/test/logging_test.go [hunk 6]
```diff
@@ -37,7 +55,9 @@ func TestPluginsInstalled(t *testing.T) {
 		installedVersion, ok := installedPlugins[pluginName]
 		assert.Truef(t, ok, "Missing plugin %s", pluginName)
 		if pluginVersion != "" {
-			assert.Equalf(t, pluginVersion, installedVersion,
+			// Getting only the version in the format 1.5 instead of 1.5.6, because installation automatically
+			// fetches the latest subversion
+			assert.Equalf(t, pluginVersion, installedVersion[:3],
```
Comment (sebsoto) [inline tools/ansible/tasks/logging/test/logging_test.go]: Not all versions will be semver so this wont work

File: tools/ansible/tasks/logging/test/logging_test.go (hunk 1 above)
Comment (suhanime) [reply tools/ansible/tasks/logging/test/logging_test.go]: os.exec for some reason isn't able to map `ruby` to the executable path of Ruby (In this case it is C:\Ruby25-x64\bin\ruby.exe), even though the bin path has been added to `PATH` environment variable as a part of installation.
os.exec only works for fully qualified path in case of `ruby --version`, otherwise it throws file-not-found error. So the fully qualified path is explicitly mentioned here.

File: tools/ansible/tasks/logging/test/logging_test.go (hunk 5 above)
Comment (suhanime) [reply tools/ansible/tasks/logging/test/logging_test.go]: No need - I will make both `rubyVersion` and `rubyExecutablePath` variables local to `func TestRubyInstalled` in the upcoming commit

File: tools/ansible/tasks/logging/test/logging_test.go (hunk 4 above)
Comment (suhanime) [reply tools/ansible/tasks/logging/test/logging_test.go]: We are only checking the version for Fluentd plugin because the playbook has an option to specify the version of Fluentd installed. We are not checking the versions for the other plugins

File: tools/ansible/tasks/logging/test/logging_test.go [hunk 7]
```diff
@@ -1,18 +1,26 @@
...
 	"os/exec"
 	"strings"
 	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
+// Pining down version of Ruby Devkit to 2.5, 64 bit version
+// Ruby installation path kept default C:\
```
Comment (aravindhp) [inline tools/ansible/tasks/logging/test/logging_test.go]: Nit: not sure what you mean by this.

File: tools/ansible/tasks/logging/test/logging_test.go [hunk 8]
```diff
@@ -25,7 +33,17 @@ var requiredPluginList = map[string]string{
 	"fluent-plugin-prometheus":                 "",
 }
 
-// TestPluginsInstalled tests that the fluentd plugins were all installed correctly.
+// TestRubyInstalled tests if Ruby is installed correctly
+// This test should be run on a node after the logging playbook has been run against it
+// exec.Command requires fully qualified path for ruby.exe
```
Comment (aravindhp) [inline tools/ansible/tasks/logging/test/logging_test.go]: This comment should be against the `exec.Command` line

File: tools/ansible/tasks/logging/test/logging_test.go (hunk 3 above)
Comment (suhanime) [reply tools/ansible/tasks/logging/test/logging_test.go]: I'll add the example in a comment here

File: tools/ansible/tasks/logging/test/logging_test.go (hunk 2 above)
Comment (suhanime) [reply tools/ansible/tasks/logging/test/logging_test.go]: I meant here that even if we mention the installation version of fluentd as `1.5.1` while running the playbook, the installation automatically fetches the latest subversion, which in this case is `1.5.2`. So I have only added the check for the version, not the subversion (i.e. I'm only match `1.5` instead of `1.5.2`.) I can reword the comment so it explains it better.

File: tools/ansible/tasks/logging/test/logging_test.go (hunk 6 above)
Comment (suhanime) [reply tools/ansible/tasks/logging/test/logging_test.go]: Correct. I'll mae the assert here only specific to Fluentd and call it out in the comments.

File: tools/ansible/tasks/logging/test/logging_test.go (hunk 1 above)
Comment (aravindhp) [reply tools/ansible/tasks/logging/test/logging_test.go]: That is because you are not running it in a shell like cmd or powershell. Only then will the PATH env get picked up.

Comment (suhanime) [reply tools/ansible/tasks/logging/test/logging_test.go]: I'm not sure that is the only reason, considering the command `exec.Command("fluent-gem", "list").Output()` in `func TestPluginsInstalled` seems to be working but `exec.Command("ruby", "--version").Output()` fails.
The `PATH` variable on the instance only reflects the bin path of Ruby, not the fluent-gem.
The error for `ruby --version` is exit status 2, which according to https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499- is `ERROR_FILE_NOT_FOUND` or `The system cannot find the file specified.`, but stdout and stderr are returned empty.
Consequently I propose to keep the executable path as fully qualified path for the unit test

File: tools/ansible/tasks/logging/test/logging_test.go (hunk 7 above)
Comment (suhanime) [reply tools/ansible/tasks/logging/test/logging_test.go]: There is an option to specify installation directory in the playbook, so I have mentioned the installation path I am assuming here

File: tools/ansible/tasks/logging/test/logging_test.go [hunk 9]
```diff
@@ -37,8 +55,16 @@ func TestPluginsInstalled(t *testing.T) {
 		installedVersion, ok := installedPlugins[pluginName]
 		assert.Truef(t, ok, "Missing plugin %s", pluginName)
 		if pluginVersion != "" {
-			assert.Equalf(t, pluginVersion, installedVersion,
-				"%s expected version: %s, got: %s", pluginName, pluginVersion, installedVersion)
+			// Test version specific to fluentd
+			if pluginName == "fluentd" {
```
Comment (sebsoto) [inline tools/ansible/tasks/logging/test/logging_test.go]: Having to do this check is unfortunate but understandable.

File: tools/ansible/tasks/logging/test/logging_test.go [hunk 10]
```diff
@@ -37,8 +55,16 @@ func TestPluginsInstalled(t *testing.T) {
...
+				// Matching only the version (ex 1.5), not the subversion (ex 1.5.1) of Fluentd because
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/logging/test/logging_test.go]: Are you sure that sub-version and patch number cannot be pinned?

File: tools/ansible/tasks/logging/test/logging_test.go [hunk 11]
```diff
@@ -25,7 +28,22 @@ var requiredPluginList = map[string]string{
 	"fluent-plugin-prometheus":                 "",
 }
 
-// TestPluginsInstalled tests that the fluentd plugins were all installed correctly.
+// TestRubyInstalled tests if Ruby is installed correctly
+// This test should be run on a node after the logging playbook has been run against it
+func TestRubyInstalled(t *testing.T) {
+	// Pinning down version of Ruby Devkit to 2.5.7, 64 bit version
+	// Ruby installation path assumed to be default
+	var rubyExecutablePath = "C:\\Ruby25-x64\\bin\\ruby.exe"
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/logging/test/logging_test.go]: So, this test need to be changed every time we upgrade the ruby version in the ansible playbook?

File: tools/ansible/tasks/logging/test/logging_test.go (hunk 10 above)
Comment (suhanime) [reply tools/ansible/tasks/logging/test/logging_test.go]: Yes - I tested it again with `gem install fluentd -v '~> 1.5.1'` and it resulted in `fluentd (1.5.2 x64-mingw32)` being installed.

File: tools/ansible/tasks/logging/test/logging_test.go (hunk 11 above)
Comment (suhanime) [reply tools/ansible/tasks/logging/test/logging_test.go]: We are providing the path to download the Ruby Installer as a variable while running the playbook.
This test needs to be changed only when the way Ruby is installed changes for some specific version.

File: tools/ansible/tasks/logging/test/logging_test.go (hunk 1 above)
Comment (aravindhp) [reply tools/ansible/tasks/logging/test/logging_test.go]: This sounds very strange and needs further digging into maybe on a hack day

File: tools/ansible/tasks/logging/test/logging_test.go (hunk 3 above)
Comment (aravindhp) [reply tools/ansible/tasks/logging/test/logging_test.go]: I think what @ravisantoshgudimetla was asking was for an example of the output `ruby --version`. Including that will help us understand how this works.

**Comment by aravindhp** [comment]
@suhanime please address https://github.com/openshift/windows-machine-config-operator/pull/80#pullrequestreview-313571929, squash the commits and I will remove the hold.

File: tools/ansible/tasks/logging/test/logging_test.go (hunk 10 above)
Comment (suhanime) [reply tools/ansible/tasks/logging/test/logging_test.go]: Can be fixed with `gem install fluentd -v 1.5.1`. Since this requires changing `install_fluentd` playbook as well, a story created at https://jira.coreos.com/browse/WINC-133.
